src/GraphQL/Validation.hs

```diff
@@ -0,0 +1,103 @@
+module GraphQL.Validation
+  ( ValidationError(..)
+  , ValidDocument
+  , validate
+  , getErrors
+  ) where
+
+import Protolude
+
+import qualified Data.GraphQL.AST as AST
+
+{-
+enum DogCommand { SIT, DOWN, HEEL }
+
+type Dog implements Pet {
+  name: String!
+  nickname: String
+  barkVolume: Int
+  doesKnowCommand(dogCommand: DogCommand!): Boolean!
+  isHousetrained(atOtherHomes: Boolean): Boolean!
+  owner: Human
+}
+
+interface Sentient {
+  name: String!
+}
+
+interface Pet {
+  name: String!
+}
+
+type Alien implements Sentient {
+  name: String!
+  homePlanet: String
+}
+
+type Human implements Sentient {
+  name: String!
+}
+
+enum CatCommand { JUMP }
+
+type Cat implements Pet {
+  name: String!
+  nickname: String
+  doesKnowCommand(catCommand: CatCommand!): Boolean!
+  meowVolume: Int
+}
+
+union CatOrDog = Cat | Dog
+union DogOrHuman = Dog | Human
+union HumanOrAlien = Human | Alien
+
+type QueryRoot {
+  dog: Dog
+}
+-}
+
+newtype ValidDocument = Valid AST.Document deriving (Eq, Show)
+
+validate :: Alternative m => AST.Document -> m ValidDocument
```

Comment:
what's the purpose of validate? Make sure the query is valid? Matches the types?